### PR TITLE
decide push down conditions during plan building phase.

### PR DESCRIPTION
**Pull request by hanfei1991**
@shenli @coocood @zimulala @tiancaiamao @XuHuaiyu PTAL

### CODE REVIEW

File: plan/physical_plans.go [hunk 1]
```diff
@@ -40,6 +41,7 @@ type PhysicalIndexScan struct {
 
 	accessEqualCount int
 	AccessCondition  []expression.Expression
+	ConditionPBExpr  *tipb.Expr
```
Comment (shenli) [inline plan/physical_plans.go]: Add comment for the newly added attribute.

File: plan/expr_to_pb.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+func expressionsToPB(exprs []expression.Expression, client kv.Client) (pbExpr *tipb.Expr, remained []expression.Expression, err error) {
```
Comment (shenli) [inline plan/expr_to_pb.go]: Should we remove the code related to expressionToPB in executor package?

Comment (hanfei1991) [inline plan/expr_to_pb.go]: yes, we should. But currently we can't, because we still need convert agg function to pb.

Comment (shenli) [inline plan/expr_to_pb.go]: Got it.

**Comment by shenli** [comment]
Rest LGTM
@coocood @zimulala @tiancaiamao PTAL

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

File: plan/physical_plan_builder.go [hunk 3]
```diff
@@ -86,14 +87,34 @@ func (p *DataSource) handleTableScan(prop requiredProperty) (*physicalPlanInfo,
...
 		for _, cond := range sel.Conditions {
 			conds = append(conds, cond.DeepCopy())
 		}
 		ts.AccessCondition, newSel.Conditions = detachTableScanConditions(conds, table)
-		err := buildNewTableRange(ts)
+		if txn != nil {
+			client := txn.GetClient()
+			var memDB bool
+			switch p.DBName.L {
+			case "information_schema", "performance_schema":
```
Comment (zimulala) [inline plan/physical_plan_builder.go]: Why not use `if` ?

Comment (hanfei1991) [inline plan/physical_plan_builder.go]: This pr has been merge ... , btw , this code is ported from executor.go

Comment (hanfei1991) [inline plan/physical_plan_builder.go]: welcome to review #1730 and leave your suggestion :)

**Pull request merged**